packages/charts/src/chart_types/heatmap/state/selectors/picked_shapes.ts

```diff
@@ -10,17 +10,21 @@ import { GlobalChartState } from '../../../../state/chart_state';
 import { createCustomCachedSelector } from '../../../../state/create_selector';
 import { Cell, TextBox } from '../../layout/types/viewmodel_types';
 import { geometries } from './geometries';
+import { getGridHeightParamsSelector } from './get_grid_full_height';
 
 function getCurrentPointerPosition(state: GlobalChartState) {
   return state.interactions.pointer.current.position;
 }
 
 /** @internal */
 export const getPickedShapes = createCustomCachedSelector(
-  [geometries, getCurrentPointerPosition],
-  (geoms, pointerPosition): Cell[] | TextBox => {
+  [geometries, getCurrentPointerPosition, getGridHeightParamsSelector],
+  (geoms, pointerPosition, gridParams): Cell[] | TextBox => {
     const picker = geoms.pickQuads;
     const { x, y } = pointerPosition;
-    return picker(x, y);
+    const arrayPicker = picker(x, y) as Cell[];
+    return Array.isArray(picker(x, y))
+      ? arrayPicker.filter(({ y }) => y < gridParams.gridCellHeight * gridParams.pageSize)
+      : picker(x, y);
```

Comment:
My suggestion is: 
- first of all call the `picker` function only one time
- don't cast it to `Cell[]` because it is not required in the code
- the ternary check `isArray` already allows TS and also JS to define if the pickedData is an array of Cell or a TextBox 
 

```suggestion
    const pickedData = picker(x, y);
    return Array.isArray(pickedData)
      ? pickedData.filter(({ y }) => y < gridParams.gridCellHeight * gridParams.pageSize)
      : pickedData;
```

Reply:
Thank you! 6785db6 for changes and added the story changes into this commit as well